Add gradients for FilledExtrapolations

SomTambe opened the pull request:
Solves #438 and maybe #434. 
```julia
julia> x = rand(RGB,3,3)
3×3 Array{RGB{Float64},2} with eltype RGB{Float64}:
 RGB{Float64}(0.30544,0.136343,0.857564)    RGB{Float64}(0.34268,0.0164714,0.261636)  RGB{Float64}(0.0954148,0.197293,0.928313)
 RGB{Float64}(0.242728,0.960734,0.249229)   RGB{Float64}(0.414405,0.771848,0.286606)  RGB{Float64}(0.23209,0.623632,0.0502413)
 RGB{Float64}(0.921937,0.0466274,0.816069)  RGB{Float64}(0.781229,0.915429,0.926183)  RGB{Float64}(0.679568,0.737393,0.702759)

julia> itp2 = interpolate(x, BSpline(Linear()))
3×3 interpolate(::Array{RGB{Float64},2}, BSpline(Linear())) with element type RGB{Float64}:
 RGB{Float64}(0.30544,0.136343,0.857564)    RGB{Float64}(0.34268,0.0164714,0.261636)  RGB{Float64}(0.0954148,0.197293,0.928313)
 RGB{Float64}(0.242728,0.960734,0.249229)   RGB{Float64}(0.414405,0.771848,0.286606)  RGB{Float64}(0.23209,0.623632,0.0502413)
 RGB{Float64}(0.921937,0.0466274,0.816069)  RGB{Float64}(0.781229,0.915429,0.926183)  RGB{Float64}(0.679568,0.737393,0.702759)

julia> etp2 = extrapolate(itp2, RGB(.0,.0,.0))
3×3 extrapolate(interpolate(::Array{RGB{Float64},2}, BSpline(Linear())), RGB{Float64}(0.0,0.0,0.0)) with element type RGB{Float64}:
 RGB{Float64}(0.30544,0.136343,0.857564)    RGB{Float64}(0.34268,0.0164714,0.261636)  RGB{Float64}(0.0954148,0.197293,0.928313)
 RGB{Float64}(0.242728,0.960734,0.249229)   RGB{Float64}(0.414405,0.771848,0.286606)  RGB{Float64}(0.23209,0.623632,0.0502413)
 RGB{Float64}(0.921937,0.0466274,0.816069)  RGB{Float64}(0.781229,0.915429,0.926183)  RGB{Float64}(0.679568,0.737393,0.702759)

julia> Interpolations.gradient(etp2,3,3)
2-element StaticArrays.SVector{2, RGB{Float64}} with indices SOneTo(2):
 RGB{Float64}(0.4474772688335722,0.11376103826212525,0.6525178769801145)
 RGB{Float64}(-0.10166125750925725,-0.17803622073108794,-0.2234233554449263)

julia> Interpolations.gradient(etp2,4,4)
2-element StaticArrays.SVector{2, RGB{Float64}} with indices SOneTo(2):
 RGB{Float64}(0.0,0.0,0.0)
 RGB{Float64}(0.0,0.0,0.0)
```

SomTambe commented:
cc. @johnnychen94 @DhairyaLGandhi

johnnychen94 reviewed (commented):
this function is better moved to filled.jl

and most importantly, tests are needed.

mkitti commented:
Let me know when this is ready for review or needs a CI run

SomTambe commented:
For what types of constants should I test for here? The normal `FloatType` only right?

DhairyaLGandhi commented on src/extrapolation/filled.jl:
```diff
@@ -54,6 +54,16 @@ end
 
+@inline function Interpolations.gradient(etp::FilledExtrapolation{T}, x::Vararg{Number,N}) where {T,N}
+    if checkbounds(Bool, etp.itp, x...)
+        gradient(etp.itp, x...)
```
`parent(etp)` instead of `etp.itp` to be symmetric with the other `gradient` function?

SomTambe replied:
Okay that sounds fine.

johnnychen94 reviewed (commented):
> For what types of constants should I test for here? The normal FloatType only right?

FloatType should be sufficient enough. But since we're targeting DiffImages, we can add Colorant types as well (and also to `Flat()`/`Line` types). More test cases are always appreciated and help catches potential issues.

Please also add `@inferred` test to make sure type stability.

SomTambe commented:
I think we should do one thing - add the `FloatType` ones here, and the `Colorant` types in the DiffImages to ensure that any of these aren't breaking for our use. What do you think @johnnychen94 ?

johnnychen94 commented:
> I think we should do one thing - add the FloatType ones here, and the Colorant types in the DiffImages to ensure that any of these aren't breaking for our use. What do you think @johnnychen94 ?

You won't test `Interpolations.gradient` in DIffImages, will you? The purpose of unit test is to break large things down into small parts, and test their basic functionality; only testing if the network forward/backward runs "normally" in its best scenario only makes sure things run correctly as an entire project. It doesn't tell you whether every single part of it runs normally, and it won't tell you which part should be checked when it goes wrong. In this sense, the unit test helps for future debugging. (You can't guarantee that you're still the maintainer three years later, and unit test helps future maintainers to understand things.)

It is often the case that people are too lazy to write tests, I understand it. And as I've said, more test cases are always appreciated and help catches potential issues.

SomTambe commented:
@mkitti I've added the tests, I think we are good to go for the CI.

mkitti commented:
Looks like we have some test issues based on type inference. The problem actually may not be in the new code, but older code. I'll have to take a closer look.

johnnychen94 commented on test/runtests.jl:
```diff
@@ -1,3 +1,4 @@
+using Base: Float64, Float32
```
This is unnecessary.

johnnychen94 commented on test/gradient.jl:
```diff
@@ -199,4 +199,20 @@ using Test, Interpolations, DualNumbers, LinearAlgebra
+        dims = [1, 2, 3]
+        for dt in dtypes
+            for dm in dims
+                itp = interpolate(rand(dt, ntuple(_ -> 5, dm)...), BSpline(Linear()))
+                etp = extrapolate(itp, rand(dt))
```
Let's also test in-range gradients are computed correctly.

```julia
@test @inferred(Interpolations.gradient(etp, 3, 3)) == @inferred(Interpolations.gradient(itp, 3, 3))
```

johnnychen94 commented on src/extrapolation/filled.jl:
```diff
@@ -54,6 +54,17 @@ end
 
+@inline function Interpolations.gradient(etp::FilledExtrapolation{T}, x::Vararg{Number,N}) where {T,N}
+    itp = parent(etp)
+    if checkbounds(Bool, itp, x...)
+        gradient(itp, x...)
+    else
+        e = eltype(itp)
+        ptype = promote_type(e, typeof(etp.fillvalue))
+        return SVector{ndims(etp), ptype}(zeros(ptype, ndims(etp)))
+    end
```
The type instability (`@inferred` failures) comes from the fact that this if-else block might return different types, hence @SomTambe you might need to double-check if `ptype` is appropriate. See https://docs.julialang.org/en/v1/manual/performance-tips/#Write-%22type-stable%22-functions for more explanation.

cc: @mkitti

johnnychen94 reviewed (commented):
It's not required for the scope of this PR but it is appreciated to also add the same test cases for extrapolation with other schemes; they're currently missing as far as I can tell.

Once the test passes I believe it is good to merge.

SomTambe commented:
Oh I just realized I made a really bad error while making the tests. Please wait until I update them.

johnnychen94 commented on test/gradient.jl:
```diff
@@ -206,12 +206,26 @@ using Test, Interpolations, DualNumbers, LinearAlgebra
-                @test all(iszero, @inferred(Interpolations.gradient(etp, -6., -6.)))
+                itp_n = interpolate(rand(dt, ntuple(_ -> 5, dm)...), BSpline(Linear()))
+                etp_n = extrapolate(itp_n, rand(dt))
+                if dm==2
+                    @test all(iszero, @inferred(Interpolations.gradient(etp_n, 6., 6.)))
+                    @test all(iszero, @inferred(Interpolations.gradient(etp_n, -6., 6.)))
+                    @test all(iszero, @inferred(Interpolations.gradient(etp_n, 6., -6.)))
+                    @test all(iszero, @inferred(Interpolations.gradient(etp_n, -6., -6.)))
+                elseif dm==1
+                    @test all(iszero, @inferred(Interpolations.gradient(etp_n, 6.)))
+                    @test all(iszero, @inferred(Interpolations.gradient(etp_n, -6.)))
+                else
+                    @test all(iszero, @inferred(Interpolations.gradient(etp_n, 6., 6., 6.)))
+                    @test all(iszero, @inferred(Interpolations.gradient(etp_n, -6., 6., 6.)))
+                    @test all(iszero, @inferred(Interpolations.gradient(etp_n, 6., -6., 6.)))
+                    @test all(iszero, @inferred(Interpolations.gradient(etp_n, -6., -6., 6.)))
+                    @test all(iszero, @inferred(Interpolations.gradient(etp_n, 6., 6., -6.)))
+                    @test all(iszero, @inferred(Interpolations.gradient(etp_n, -6., 6., -6.)))
+                    @test all(iszero, @inferred(Interpolations.gradient(etp_n, 6., -6., -6.)))
+                    @test all(iszero, @inferred(Interpolations.gradient(etp_n, -6., -6., -6.)))
+                end
```
These two seem sufficient to me.

```julia
for x in [ntuple(_->0, dm), ntuple(_->6, dm)]
```

SomTambe replied:
But this will fall short for the 2-dim and 3-dim cases where we have multiple quadrants right? It's better to get it checked for every quadrant right? That is why I hardcoded them.

johnnychen94 replied:
Yeah I get your points but it's not really necessary. We only need to choose _some_ out of range points and test it. This is a more programmable way and is more friendly for future maintainers.

SomTambe replied:
Ok I have changed it.

DhairyaLGandhi commented on src/extrapolation/filled.jl:
```diff
@@ -60,7 +60,7 @@ end
     else
         e = eltype(itp)
-        ptype = promote_type(e, typeof(etp.fillvalue))
+        ptype = promote_type(e, mapreduce(typeof, promote_type, x))
```
Can't we let Julia's type system figure out the types? Arebts things like `itp` already parameterized?

johnnychen94 replied:
Not really, this is because interpolations gradient use `x`(the indices) to generate `δx` and thus the gradient output, to keep type stability we need to keep consistent to interpolation gradient internal.

johnnychen94 commented on src/extrapolation/filled.jl:
```diff
@@ -54,6 +54,17 @@ end
+    if checkbounds(Bool, itp, x...)
+        gradient(itp, x...)
+    else
+        e = eltype(itp)
+        ptype = promote_type(e, mapreduce(typeof, promote_type, x))
+        return SVector{ndims(etp), ptype}(zeros(ptype, ndims(etp)))
```
`zeros(ptype, ndims(etp))` is unnecessary intermediate memory allocation, let's use tuple here.

```suggestion
        ptype = promote_type(eltype(itp), map(typeof, x)...)
        return SVector{ndims(etp), ptype}(ntuple(_->zero(ptype), ndims(etp)))
```

```julia
julia> etp = extrapolate(interpolate([1, 2, 3], BSpline(Linear())), 0);

# before
julia> @btime Interpolations.gradient($etp, -0.5f0);
  35.995 ns (1 allocation: 96 bytes)
# after
julia> @btime Interpolations.gradient($etp, -0.5f0);
  0.038 ns (0 allocations: 0 bytes)
```

SomTambe commented:
@mkitti Could you run CI on this? I have corrected type instability and also added more tests for the other boundary conditions as there weren't any written for them.

mkitti commented:
Looks good. Merging. Thank you.

SomTambe commented:
The code is breaking for these types - 
```julia
julia> x = rand(RGB,3,3);

julia> itp2 = interpolate(x, BSpline(Linear()));

julia> etp2 = extrapolate(itp2, RGB(.0,.0,.0));

julia> Interpolations.gradient(etp2,3,3)
2-element SVector{2, RGB{Float64}} with indices SOneTo(2):
 RGB{Float64}(0.43457141859445714,0.1577128433792203,0.02876792599552047)
 RGB{Float64}(0.19412021621096764,-0.5215101620517046,0.36728033563097706)

julia> Interpolations.gradient(etp2,4,4) # wrong, should be of types RGB{Float64}
2-element SVector{2, Any} with indices SOneTo(2):
 0.0
 0.0
```
This is coming from the lines [62](https://github.com/JuliaMath/Interpolations.jl/blob/master/src/extrapolation/filled.jl#L62) in `Filled.jl` - 
```julia
...
ptype = promote_type(eltype(itp), map(typeof, x)...)
...
```
This renders `ptype` to be `Any` instead of being `RGB{Float64}`, which is further causing problems in gradients to propogate in DiffImages.jl. 
I wonder how I missed this.
cc. @johnnychen94 @DhairyaLGandhi

johnnychen94 commented:
> FloatType should be sufficient enough. But since we're targeting DiffImages, we can add Colorant types as well (and also to Flat()/Line types). More test cases are always appreciated and help catches potential issues.

A typical case that more test matters.

This is because Numbers can't be promoted to Colorants unless specified:

```julia
julia> promote_type(RGB{Float32}, Int)
Any # We would expect this to be RGB{Float32}
```

The reason we use `promote_type` is to mimic the output type of `gradient` for *Interpolation for type stability. Since this trick doesn't work, we might need to look deeper into the codebase to see how we can get that.

> Can't we let Julia's type system figure out the types?

Back to the comment, @DhairyaLGandhi do you mean to use `Base._return_type` to get the inferred output type?

DhairyaLGandhi commented:
> did you mean to use Base._return_type to get the inferred output type?

Yes, or change the method signature to use `Vararg{S} where S`. Are there common cases where indices would not be the same type? I would consider that a bug if anything.
